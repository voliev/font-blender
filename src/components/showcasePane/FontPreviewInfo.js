import React from "react";
import PropTypes from "prop-types";
import "./styles/FontPreviewInfo.css";

const FontPreviewInfo = ({ family, category, subsets }) => {
  const style = {
    fontFamily: `'${family}', ${category}`
  };

  return (
    <div className="FontPreviewInfo">
      <h2 className="heading" style={style}>
        {family}
      </h2>
      <div className="category">
        <h3 className="heading">Category</h3>
        <p className="category-name">{category}</p>
      </div>
      <div className="charsets">
        <h3 className="heading">Charsets</h3>
        <ul className="charsets-list">
          {subsets.map(subset => (
            <li key={subset}>{subset}</li>
          ))}
        </ul>
      </div>
    </div>
  );
};

FontPreviewInfo.propTypes = {
  family: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  subsets: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired
};

export default FontPreviewInfo;
