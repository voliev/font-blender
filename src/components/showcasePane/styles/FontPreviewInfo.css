.FontPreviewInfo .heading {
  border-bottom: 1px solid var(--light-grey);
  padding-bottom: var(--space-md);
}

.FontPreviewInfo .category-name,
.FontPreviewInfo .charsets-list {
  color: var(--dark-grey);
}

.FontPreviewInfo .charsets-list {
  padding-left: var(--space-md);
}
