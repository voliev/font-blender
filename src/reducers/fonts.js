import { combineReducers } from "redux";
import {
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fontsDataStatus = (state = "", action) => {
  switch (action.type) {
    case FETCH_FONTS_DATA:
      return "fetching";
    case FETCH_FONTS_DATA_SUCCESS:
      return "success";
    case FETCH_FONTS_DATA_FAILURE:
      return "failure";
    default:
      return state;
  }
};

const initialState = {};

export const all = (state = initialState, action) => {
  switch (action.type) {
    case SET_FONTS:
      return action.fonts;
    case UPDATE_FONT:
      return {
        ...state,
        [action.family]: {
          ...state[action.family],
          ...action.updates
        }
      };
    default:
      return state;
  }
};

export default combineReducers({
  all,
  fontsDataStatus
});

export const getFonts = state => state.fonts.all;
export const getFont = (state, family) => state.fonts.all[family];
export const getFontsDataStatus = state => state.fonts.fontsDataStatus;
