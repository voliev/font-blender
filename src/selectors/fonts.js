import isEqual from "lodash.isequal";

export default function() {
  const memo = {
    filters: {},
    result: []
  };

  return function(fonts, filters) {
    const families = Object.keys(fonts);

    if (families.length === 0) {
      memo.filters = { ...filters };
      memo.result = [];

      return [];
    }

    if (
      (memo.result.length === 0 && families.length > 0) ||
      !isEqual(memo.filters, filters)
    ) {
      memo.filters = { ...filters };

      const { title, categories, sorting } = filters;

      const filteredFonts = families.filter(family => {
        const { category } = fonts[family];

        const titleMatch = family
          .toLowerCase()
          .includes(title.trim().toLowerCase());
        const categoryMatch = categories[category];

        return titleMatch && categoryMatch;
      });

      if (sorting.param === "alpha") {
        const sortedFonts = filteredFonts.sort((a, b) =>
          a.localeCompare(b, {}, { caseFirst: "upper" })
        );

        memo.result =
          sorting.order === "asc" ? sortedFonts : sortedFonts.reverse();
      } else {
        memo.result =
          sorting.order === "asc" ? filteredFonts : filteredFonts.reverse();
      }
    }

    return memo.result;
  };
}
