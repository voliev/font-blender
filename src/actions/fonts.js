import WebFont from "webfontloader";
import { getFonts } from "../reducers/fonts";
import {
  GOOGLE_FONTS_API_ENDPOINT,
  FETCH_FONTS_DATA,
  FETCH_FONTS_DATA_FAILURE,
  FETCH_FONTS_DATA_SUCCESS,
  SET_FONTS,
  UPDATE_FONT
} from "../constants";

export const fetchFontsData = () => ({
  type: FETCH_FONTS_DATA
});

export const fetchFontsDataSuccess = () => ({
  type: FETCH_FONTS_DATA_SUCCESS
});

export const fetchFontsDataFailure = () => ({
  type: FETCH_FONTS_DATA_FAILURE
});

export const setFonts = fonts => ({
  type: SET_FONTS,
  fonts
});

export const startFetchFontsData = () => {
  return async dispatch => {
    const apiRequestURL = `${GOOGLE_FONTS_API_ENDPOINT}?key=${
      process.env.REACT_APP_GOOGLE_FONTS_API_KEY
    }&sort=popularity`;

    dispatch(fetchFontsData());

    try {
      const res = await fetch(apiRequestURL);
      const { items } = await res.json();
      dispatch(fetchFontsDataSuccess());

      const fonts = items.reduce((result, font) => {
        result[font.family] = {
          family: font.family,
          category: font.category,
          variants: font.variants,
          subsets: font.subsets,
          fetching: false,
          active: false,
          requested: false,
          prefetched: false,
          fetched: false
        };

        return result;
      }, {});

      return dispatch(setFonts(fonts));
    } catch (e) {
      dispatch(fetchFontsDataFailure());
      // TODO: Eject and replace service worker with custom one.
      //       Implement fonts' json data caching.
      return dispatch(setFonts({}));
    }
  };
};

export const updateFont = (family, updates) => ({
  type: UPDATE_FONT,
  family,
  updates
});

export const startFetchFont = (family, charRange = undefined) => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];
    const fetchFull = !charRange;
    const requestString = generateRequestString(font, fetchFull);
    const options = generateRequestOptions(
      requestString,
      charRange,
      dispatch,
      state
    );

    WebFont.load(options);

    const updatedState = getState();
    const updatedFonts = getFonts(updatedState);

    // For "then" chaining in selectedFonts.js async action creator
    return Promise.resolve(updatedFonts[family]);
  };
};

export const generateRequestString = (font, fetchFull = false) => {
  const { family, subsets, variants } = font;

  if (fetchFull) {
    return `${family}:${variants.join(",")}`;
  }

  const weight = variants.includes("regular")
    ? "400"
    : parseInt(variants[0], 10).toString();
  const subset = subsets.includes("latin") ? "latin" : subsets[0];

  return `${family}:${weight}:${subset}`;
};

export const generateRequestOptions = (request, chars, dispatch) => ({
  google: {
    families: [request],
    text: chars
  },
  classes: false,
  timeout: 10000,
  fontloading: getEventListener("fontloading", dispatch, chars),
  fontactive: getEventListener("fontactive", dispatch, chars),
  fontinactive: getEventListener("fontinactive", dispatch, chars)
});

export const getEventListener = (event, dispatch, charRange) => {
  const listeners = {
    fontloading: family => {
      const updates = Object.assign(
        { fetching: true },
        charRange ? { requested: true } : {}
      );
      dispatch(updateFont(family, updates));
    },
    fontactive: family => {
      const updates = Object.assign(
        { active: true, fetching: false },
        charRange ? { prefetched: true } : { fetched: true }
      );
      dispatch(updateFont(family, updates));
    },
    fontinactive: family => {
      const updates = Object.assign(
        { fetching: false },
        charRange ? { active: false } : {}
      );
      dispatch(updateFont(family, updates));
    }
  };

  return listeners[event];
};
