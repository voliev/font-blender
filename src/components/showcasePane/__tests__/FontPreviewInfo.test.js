import React from "react";
import { shallow } from "enzyme";
import FontPreviewInfo from "../FontPreviewInfo";
import fonts from "../../../fixtures/fonts";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const font = fonts["Roboto"];
  const props = {
    family: font.family,
    category: font.category,
    subsets: font.subsets,
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: renderFn(<FontPreviewInfo {...props} />)
  };
};

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should pass corresponding style", () => {
  const { family, category, wrapper } = setup();

  expect(wrapper.find("h2").prop("style")).toEqual({
    fontFamily: `'${family}', ${category}`
  });
});

it("should render charsets list", () => {
  const { subsets, wrapper } = setup();

  wrapper
    .find("li")
    .forEach((node, idx) => expect(node.text()).toBe(subsets[idx]));
});

it("should render category name", () => {
  const { category, wrapper } = setup();

  expect(wrapper.find(".category-name").text()).toBe(category);
});
