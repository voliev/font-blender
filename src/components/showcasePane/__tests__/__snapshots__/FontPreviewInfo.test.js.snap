// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<div
  className="FontPreviewInfo"
>
  <h2
    className="heading"
    style={
      Object {
        "fontFamily": "'Roboto', sans-serif",
      }
    }
  >
    Roboto
  </h2>
  <div
    className="category"
  >
    <h3
      className="heading"
    >
      Category
    </h3>
    <p
      className="category-name"
    >
      sans-serif
    </p>
  </div>
  <div
    className="charsets"
  >
    <h3
      className="heading"
    >
      Charsets
    </h3>
    <ul
      className="charsets-list"
    >
      <li
        key="greek"
      >
        greek
      </li>
      <li
        key="vietnamese"
      >
        vietnamese
      </li>
      <li
        key="latin-ext"
      >
        latin-ext
      </li>
      <li
        key="cyrillic-ext"
      >
        cyrillic-ext
      </li>
      <li
        key="greek-ext"
      >
        greek-ext
      </li>
      <li
        key="latin"
      >
        latin
      </li>
      <li
        key="cyrillic"
      >
        cyrillic
      </li>
    </ul>
  </div>
</div>
`;
